### PR TITLE
deregister events

### PR DESCRIPTION
**Pull request by lee-orr**
# Objective

Add ability to de-register events from the EventRegistry (and the associated World).

The initial reasoning relates to retaining support for Event hot reloading in `dexterous_developer`.

## Solution

Add a `deregister_events<T: Event>(&mut world)` method to the `EventRegistry` struct.

## Testing

Added an automated test that verifies the event registry adds and removes `Events<T>` from the world.

### CODE REVIEW

**Review by alice-i-cecile — APPROVED** [review]
Great, I like this. There's no reason to expose this on `App`: this is extremely niche.

No concerns about shipping this in 0.14: this is purely additive, simple and has a test.

File: crates/bevy_ecs/src/event/registry.rs [hunk 1]
```diff
@@ -77,4 +77,26 @@ impl EventRegistry {
...
+                    Some(i)
+                } else {
+                    None
+                }
+            })
+        else {
+            return;
+        };
+        registry.event_updates.remove(index);
+    }
```
Comment (MiniaczQ) [inline crates/bevy_ecs/src/event/registry.rs:101]: Can't we use `retain`?

File: crates/bevy_ecs/src/event/registry.rs [hunk 2]
```diff
@@ -77,4 +77,26 @@ impl EventRegistry {
             }
         }
     }
+
+    /// Removes an event from the registry and world
+    pub fn deregister_events<T: Event>(world: &mut World) {
```
Comment (MiniaczQ) [inline crates/bevy_ecs/src/event/registry.rs:82]: Word of warning of how this will affect existing `EventReader`s
We can do this in another PR, but it's a pretty important side effect

**Review by MiniaczQ — APPROVED** [review]
Looks good

File: crates/bevy_ecs/src/event/registry.rs (hunk 2 above)
Comment (MiniaczQ) [reply crates/bevy_ecs/src/event/registry.rs:82]: Let me clarify, if we remove the `Events<E>`, running a system with `EventReader` will obviously crash,
the question is what if we re-add the same `Events<E>` back, how will that affect the reader

Comment (lee-orr) [reply crates/bevy_ecs/src/event/registry.rs:82]: I think I'll adjust it so this doesn't remove `Events<E>` - ATM it'll cause the count to re-set, but since the readers aren't re-set (and I'm not sure if there's a way to remove or reset all `Local<'s, ManualEventReader<E>>` resources) re-introducing the Events<E> will lead to it missing all events until the count reaches the latest count a given reader saw.

File: crates/bevy_ecs/src/event/registry.rs (hunk 1 above)
Comment (lee-orr) [reply crates/bevy_ecs/src/event/registry.rs:101]: We can! I was just unaware of it, and it's name wasn't clear to me from intellisense.

File: crates/bevy_ecs/src/event/registry.rs (hunk 2 above)
Comment (lee-orr) [reply crates/bevy_ecs/src/event/registry.rs:82]: Actually - changed my mind here.

Since you'd have to remove the systems anyway to prevent crashes, the Local is unlikely to be retained. I do have a potential way to prevent both the crashes & this issue in mind (generational indexes), but think that might be a bit beyond the scope here.

Comment (MiniaczQ) [reply crates/bevy_ecs/src/event/registry.rs]: You can have run conditions or schedules that you don't run,
anyways, the solution is outside of the scope for this PR, no need to sweat it rn

File: crates/bevy_ecs/src/event/registry.rs [hunk 3]
```diff
@@ -41,7 +41,13 @@ pub enum ShouldUpdateEvents {
 }
 
 impl EventRegistry {
-    /// Registers an event type to be updated.
+    /// Registers an event type to be updated in a given [`World`]
+    ///
+    /// If no instance of the EventRegistry exists in the world, this will add one - otherwise it will use
```
Comment (BD103) [inline crates/bevy_ecs/src/event/registry.rs:46]:
```suggestion
    /// If no instance of the [`EventRegistry`] exists in the world, this will add one - otherwise it will use
```

File: crates/bevy_ecs/src/event/registry.rs [hunk 4]
```diff
@@ -77,4 +103,25 @@ impl EventRegistry {
             }
         }
     }
+
+    /// Removes an event from the world and it's associated EventRegistry
```
Comment (BD103) [inline crates/bevy_ecs/src/event/registry.rs:107]:
```suggestion
    /// Removes an event from the world and it's associated [`EventRegistry`].
```

**Review by BD103 — COMMENTED** [review]
This should fix CI!

File: crates/bevy_ecs/src/event/registry.rs [hunk 5]
```diff
@@ -59,6 +65,26 @@ impl EventRegistry {
         });
     }
 
+    /// Registers an event type from a give [`World`] to be updated bt th
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/event/registry.rs:68]: This is mangled and needs to be fixed.

File: crates/bevy_ecs/src/event/registry.rs [hunk 6]
```diff
@@ -77,4 +103,25 @@ impl EventRegistry {
...
+            .event_updates
+            .retain(|e| e.component_id != component_id);
+        world.remove_resource::<Events<T>>();
+    }
+
+    /// Removes an event from this registry and world
+    ///
+    /// In contrast, [`EventRegistry::deregister_events<T>`] will remove the events from the world's
+    /// main event registry
+    pub fn deregister_events_with_instance<T: Event>(&mut self, world: &mut World) {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/event/registry.rs:121]: The with_instance naming isn't clear to me: is there an alternative way to frame this?

Comment (lee-orr) [reply crates/bevy_ecs/src/event/registry.rs:121]: The ideal way to frame it, in my opinion, would be using something like:

- renaming the  `register_events` & `deregister_events` to `register_events_in_default_registry`
- renaming the `with_instance` variants to the current names of the main ones

But since it involves essentially boils down to changing the behaviour of the currently named items, that will be a migration pain point - and not something I'd like to do right before a release.

I've also realized that there is much more work necessary to allow some way of actually using more than 1 EventRegistry in a world. So I think I'll take those elements out and return this to the simpler solution it to begin with.

Comment (lee-orr) [reply crates/bevy_ecs/src/event/registry.rs]: I have a concept I've been mulling over, and will likely set up some kind of discussion around it in the near future that will include this.

**Review by alice-i-cecile — APPROVED** [review]
